lib/image_optim/cache.rb

```diff
@@ -0,0 +1,71 @@
+require 'digest/sha1'
+require 'fspath'
+
+class ImageOptim
+  # Handles image cache
+  class Cache
+    def initialize(image_optim, workers_by_format)
+      return unless image_optim.cache_dir
+      @cache_dir = FSPath.new(image_optim.cache_dir)
+      @cache_worker_digests = image_optim.cache_worker_digests
+      @options_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:inspect).sort.join(', ')]
+      end]
+      @bins_by_format = Hash[workers_by_format.map do |format, workers|
+        [format, workers.map(&:used_bins).flatten!.map! do |sym|
+          bin = image_optim.resolve_bin!(sym)
+          "#{bin.name}[#{bin.digest}]"
+        end.sort!.uniq.join(', ')]
+      end]
+    end
+
+    def fetch(original)
+      return block_given? ? yield : nil if @cache_dir.nil?
```

Comment:
`fetch` should not be used without block, so I think you should call `yield` without checking `block_given?`